### ISSUE TITLE
set vs. upsert

### ISSUE THREAD

**Issue by jpetso** [pmwkaa/sophia#110]
Most examples (and all examples in the documented CRUD section on the website) use `sp_set()` to write documents/rows. It seems it's able to insert new values, as well as update existing ones.

As far as I'm aware, "UPSERT" in existing databases is used to (1) get around the limitation of standard SQL that there's only "INSERT" and "UPDATE" but not "SET", and (2) to avoid race conditions caused by (1).

Sophia has `sp_set()` so from a strictly intuitive point of view, `sp_upsert()` seems confusing because why would there be two functions that do the same thing? Especially since it's not really documented.

Now, looking at the upsert example, there's the "db.test.index.upsert" callback whose function signature looks a bit like that of the [RocksDB merge operator](https://github.com/facebook/rocksdb/wiki/Merge-Operator). Is that what upsert is for? Should it be named `sp_merge()` instead to make its purpose clearer?

When is it called and what's the right way to manage the memory that apparently needs to be allocated manually, especially after `sp_set()` allocates stuff internally?

Maybe relatedly, is there a way to perform the traditional SQL semantics, i.e. pure "INSERT" (error if key exists) and/or "UPDATE" (error if key doesn't exist)?

**Comment by pmwkaa** [comment]
Upsert command (Update or Insert) allows to reduce Read-Modify-Write case to a single Write. Updates are applied during data compaction or upon read request.

So yes, Upsert is very likely the merge operator, but have a slightly different semantic.

Upsert name have been chosen because in my opinion it is semantically closer to what it really does. Upsert callback operates with one or two records at the time. Strictly saving the original insertion order. For one records: it does **insert**, create and return a *unique* record using value. For two records: it should apply user-supplied **update** logic to first record to create a third record. Upsert callback is never used with other statements like set, delete, etc.
malloc(3) is expected to be used for data returned by callback.

example: https://github.com/pmwkaa/sophia/blob/master/example/upsert.c

Additionally, upsert been designed in a way that i can be used by other DBMS in which Sophia is integrated as a storage engine. For example: http://tarantool.org/doc/book/box/box_space.html?highlight=upsert#lua-function.space_object.upsert (this operation supported by sophia engine)

**is there a way to perform the traditional SQL semantics, i.e. pure "INSERT" (error if key exists) and/or "UPDATE" (error if key doesn't exist)?**

Right now this behavior is simply implemented by doing sp_get() + sp_set(). I got some ideas regarding pure insert, replace. It could be implemented write-only and automatically handled during compaction and read times, in very likely manner as sp_delete() and upsert logic does.

**Comment by arthurprs** [comment]
Question. When some values are merged due to a get call the merged value is automatically inserted back into the database?

**Comment by pmwkaa** [comment]
No. It is a little more complicated than this. Upsert statements are also take part in the transactional visibility. This means that merged values might differ for readers, depending on moment of time when a transaction started.
